### PR TITLE
WIP: Added Firebird Support

### PR DESCRIPTION
**Pull request by kerwynrg**
Based on the project mentioned in this issue https://github.com/typeorm/typeorm/issues/2880 started by https://github.com/edu-ricardo and adapted to latest release.

EDIT:
Pending:
[ ] Create before insert trigger to simulate autoincrement values, at least primary key.
[ ] Implement correct travis tests to pass.

### CODE REVIEW

**Comment by pleerock** [comment]
Thanks for contribution!

Will you be able to maintain this driver? Can you please enable it in CI (edit ormconfig travis json) so we can check if all tests are passing using this driver.

File: src/driver/firebird/FirebirdDriver.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+            } else {
+                escapedParameters.push(value);
+                return "?";
+            }
+        }); // todo: make replace only in value statements, otherwise problems
+        return [sql, escapedParameters];
+    }
+    escape(columnName: string): string {
+        return "`" + columnName + "`";
```
Comment (Kagu-chan) [inline src/driver/firebird/FirebirdDriver.ts]:
```suggestion
        return '"' + columnName + '"';
```

Backticks does not work in firebird (we're using V3.0)

File: src/driver/firebird/FirebirdDriver.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    }
+    obtainSlaveConnection(): Promise<any> {
+        return this.connect();
+    }
+    createGeneratedMap(metadata: EntityMetadata, insertResult: any): ObjectLiteral | undefined {
+        const generatedMap = metadata.generatedColumns.reduce((map, generatedColumn) => {
+            let value: any;
+            if (generatedColumn.generationStrategy === "increment" && insertResult.insertId) {
+                value = insertResult.insertId;
+            // } else if (generatedColumn.generationStrategy === "uuid") {
```
Comment (Kagu-chan) [inline src/driver/firebird/FirebirdDriver.ts]: Better remove commented code

File: src/driver/firebird/FirebirdQueryRunner.ts [hunk 3]
```diff
@@ -0,0 +1,1651 @@
...
+    async hasSchema(schema: string): Promise<boolean> {
+        throw new Error(`Firebird driver does not support table schemas`);
+    }
+
+    /**
+     * Checks if table with the given name exist in the database.
+     */
+    async hasTable(tableOrName: Table|string): Promise<boolean> {
+        const parsedTableName = this.parseTableName(tableOrName);
+        const sql = `SELECT * FROM \`RDB$RELATIONS\` WHERE \`RDB$FLAGS\`= 1 AND \`RDB$RELATION_NAME\` = '${parsedTableName.tableName}'`;
```
Comment (Kagu-chan) [inline src/driver/firebird/FirebirdQueryRunner.ts]: Here (And as well later in the code) we have the problem that not all FB versions support backticks.

```suggestion
        const sql = `SELECT * FROM "RDB$RELATIONS" WHERE "RDB$FLAGS"= 1 AND "RDB$RELATION_NAME" = '${parsedTableName.tableName}'`;
```

**Comment by kerwynrg** [comment]
I can occationally maintain this driver, I'll be more active from now on.
How is possible to test this dialect following standards of this project? I could't get to pass tests from travis b/c fdb was not created with docker image. Any suggestions @Kagu-chan

**Comment by haroldo-ok** [comment]
Also, if you need FB dialect 1 (the PR above supports dialect 3), I have made this fork for the company I work for: https://github.com/haroldo-ok/typeorm-firebird

**Comment by haroldo-ok** [comment]
If anyone needs the fork:
https://www.npmjs.com/package/typeorm-firebird

**Comment by VZiemer** [comment]
hello, i'm interested in the project, how can i help?

**Comment by wesleyguirra** [comment]
Some update on this?

**Comment by tfiliano** [comment]
Hi, 

How can I help to put this on production?

**Comment by imnotjames** [comment]
This is still marked as WIP & has a few conflicts - is it still in progress?

**Comment by carlosscheffer** [comment]
Alguma novidade sobre isso? O

**Comment by carlosscheffer** [comment]
Any news about?

**Pull request closed without merging**